proxy/src/connection.rs

```diff
@@ -219,25 +219,20 @@ impl Future for ConditionallyUpgradeServerToTls {
         loop {
             *self = match self {
                 ConditionallyUpgradeServerToTls::Plaintext(ref mut inner) => {
-                    let r = {
-                        let inner = inner.as_mut().unwrap();
-                        try_ready!(inner.socket.read_buf(&mut inner.peek_buf));
-                        tls::conditional_accept::match_client_hello(
-                            inner.peek_buf.as_ref(), &inner.tls.identity)
-                    };
-                    match r {
+                    let poll_match = inner
+                        .as_mut()
+                        .expect("polled after ready")
+                        .poll_match_client_hello();
+
+                    match try_ready!(poll_match) {
                         tls::conditional_accept::Match::Matched => {
                             trace!("upgrading accepted connection to TLS");
-                            let inner = inner.take().expect("Polled after ready");
-                            let upgrade_to_tls = tls::Connection::accept(
-                                inner.socket, inner.peek_buf.freeze(), inner.tls.config);
-                            ConditionallyUpgradeServerToTls::UpgradeToTls(upgrade_to_tls)
+                            let upgrade = inner.take().unwrap().into();
```

Comment:
TIOLI: You use `.into()` a lot in this function, and I think `tls::UpgradeServerToTls::from(...)` might be clearer. On the other hand, it's probably also a lot more verbose, so IDK...

Reply:
For whatever reason I couldn't get it to work that way. I'll worry about it later.